File: solutions/144.binary-tree-preorder-traversal/binary-tree-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
function preorderTraversal(root) {
  if (!root) return []

  const res = []

  const dfs = node => {
    res.push(node.val)

    if (node.left) {
      dfs(node.left)
    }

    if (node.right) {
      dfs(node.right)
    }
  }

  dfs(root)

  return res
}

File: solutions/144.binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
class Solution:
    def preorderTraversal(self, root):
        ret = []
        s = [root]
        while s:
            cur = s.pop()
            if not cur:
                continue

            ret.append(cur.val)

            if cur.right:
                s.append(cur.right)

            if cur.left:
                s.append(cur.left)

        return ret

    
#     def preorderTraversal(self, root):
#         """
#         :type root: TreeNode
#         :rtype: List[int]
#         root
#         left
#         right
#         """
#         ret = []

#         def traverse(node):
#             if not node:
#                 return
#             ret.append(node.val)
#             traverse(node.left)
#             traverse(node.right)

#         traverse(root)
#         return ret
